core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -56,7 +60,9 @@ class LogManager(val logDirs: Array[File],
                  scheduler: Scheduler,
                  val brokerState: BrokerState,
                  brokerTopicStats: BrokerTopicStats,
-                 time: Time) extends Logging {
+                 val zkUtils: ZkUtils,
```

Comment:
@junrao Are you suggesting that `LogManager.handleLogDirFailure()` should invoke `ReplicaManager.handleLogDirFailure()` and `ReplicaManager` should be put in the constructor of the `LogManager`? This can be done. But I am worried that this creates circular dependency between `LogManager` and `ReplicaManager` and can make future development harder.

In addition to the worry with this Java class dependency, I also find hard to organize the methods. For example, say the broker receives a `ProduceRequest`, triggers `ReplicaManager.appendToLocalLog()`, which in turn calls `Partition.appendRecordsToLeader()` and fails with `IOException`. Note that we haven't touched `LogManger` in this path and thus `LogManger.handleLogDirFailure()` won't be called.

Now that the ReplicaManager catches an IOException, ideally it should have its own method `handleLogDirFailure()` to deal with it, e.g. remove the corresponding TopcPartition from `ReplicaManager.allPartitions` and `LogManager.logs`. Then it become pretty straightforward to have `ReplicaManager.handleLogDirFailure()` to call `LogManager.handleLogDirFailure()`. But then it seems weird for `LogManager.handleLogDirFailure()` to invoke `ReplicaManager.handleLogDirFailure()` even though we can use some trick to make it work. Do you have a solution to this circular invocation?

Currently if the IOException is triggered by user's request, both `ReplicaManager.handleLogDirFailure` and `LogManager.handleLogDirFailure()` will be called so this is OK. If IOException is triggered by e.g. `LogCleaner`, only `LogManager.handleLogDirFailure()` will be triggered and some partition will still stay in `ReplicaManager.allpartitions`. As of now I don't find this to be a problem. Because the next time any ProduceRequest or FetchRequest tries to access a partition on that offline log directory, ReplicaManager will handle the log directory failure. Also, I have updated the code so that `ReplicaManager.checkpointHighWatermarks()` will trigger `ReplicaManager.handleLogDirFailure()` if any partition in `allPartitions` is on an offline log directory. Thus the period of inconsistency will be limited. Does this address your concern?